File: python/src/transtex/parsers/chicago.py
"""Parse Chicago author-date citations."""
from __future__ import annotations

import re

from ..reference import Reference
from .shared import (
    clean_locator,
    generate_cite_key,
    normalize_pages,
    split_authors_delimited,
    strip_trailing_period,
)


def parse_chicago_citation(text: str) -> Reference:
    raw = text.strip()
    if not raw:
        raise ValueError("Empty Chicago citation string")

    match = re.match(r"^(?P<authors>.*?)\.\s+(?P<year>\d{4}|n\.d\.)\.\s+(?P<rest>.+)$", raw)
    if not match:
        raise ValueError("Chicago citation missing expected segments")
    authors_segment = match.group("authors").strip()
    year = strip_trailing_period(match.group("year").strip()) or None
    remainder = match.group("rest").strip()

    title_match = re.search(r"\"(.+?)\"\s", remainder)
    if not title_match:
        raise ValueError("Chicago citation missing title")
    title = strip_trailing_period(title_match.group(1).strip())
    after_title = remainder[title_match.end() :].strip()

    locator = None
    locator_match = re.search(r"(https?://\S+|10\.\S+)$", after_title)
    if locator_match:
        locator = locator_match.group(1).strip()
        after_title = after_title[: locator_match.start()].strip()
    after_title = after_title.rstrip(".").strip()

    journal_match = re.match(r"\*([^*]+)\*\s+([\d]+)\s*\(([^)]+)\):\s*([^\s]+)", after_title)
    journal = volume = issue = pages = None
    if journal_match:
        journal = journal_match.group(1).strip()
        volume = journal_match.group(2).strip()
        issue = journal_match.group(3).strip()
        pages = normalize_pages(journal_match.group(4).strip())

    reference = Reference(
        entry_type="article",
        cite_key=generate_cite_key([], year, title),
        title=title,
        authors=_parse_authors(authors_segment),
        journal=journal,
        volume=volume,
        issue=issue,
        pages=pages,
        year=year,
    )
    doi, url = clean_locator(locator)
    reference.doi = doi
    reference.url = url
    return reference


def _parse_authors(segment: str) -> list[str]:
    cleaned = segment.rstrip(".").strip()
    if " and " in cleaned:
        first, second = cleaned.split(" and ", 1)
        authors = [first.strip().rstrip(",")]
        authors.append(second.strip())
        return [author for author in authors if author]
    return [cleaned] if cleaned else []


__all__ = ["parse_chicago_citation"]

File: typescript/src/parsers/chicago.ts
import { Reference } from "../reference";
import { applyLocator, generateCiteKey, normalizePages, splitAuthorsDelimiters, stripTrailingPeriod } from "./shared";

export function parseChicagoCitation(text: string): Reference {
    const raw = text.trim();
    if (!raw) throw new Error("Empty Chicago citation string");

    const parts = raw.split(". ", 2);
    if (parts.length < 2) throw new Error("Chicago citation missing expected segments");
    const authorsSegment = parts[0];
    const remainder = raw.slice(authorsSegment.length + 2);

    const yearMatch = remainder.match(/^([0-9n.dND\.]+)\.\s+(.*)$/);
    if (!yearMatch) throw new Error("Chicago citation missing year segment");
    const year = stripTrailingPeriod(yearMatch[1].trim()) || undefined;
    const afterYear = yearMatch[2].trim();

    const titleMatch = afterYear.match(/"(.+?)"\s/);
    if (!titleMatch) throw new Error("Chicago citation missing title");
    const title = stripTrailingPeriod(titleMatch[1].trim());
    const afterTitle = afterYear.slice((titleMatch.index ?? 0) + titleMatch[0].length).trim();

    let locator: string | undefined;
    const locatorMatch = afterTitle.match(/(https?:\/\/\S+|10\.\S+)$/);
    let detailSegment = afterTitle;
    if (locatorMatch) {
        locator = locatorMatch[1].trim();
        detailSegment = afterTitle.slice(0, locatorMatch.index).trim();
    }

    let journal: string | undefined;
    let volume: string | undefined;
    let issue: string | undefined;
    let pages: string | undefined;

    const journalMatch = detailSegment.match(/\*([^*]+)\*\s+([\d]+)\s*\(([^)]+)\):\s*([^\s]+)/);
    if (journalMatch) {
        journal = journalMatch[1].trim();
        volume = journalMatch[2].trim();
        issue = journalMatch[3].trim();
        pages = normalizePages(journalMatch[4].trim());
    }

    const authors = parseAuthors(authorsSegment);
    const reference = new Reference({
        entryType: "article",
        citeKey: generateCiteKey(authors, year, title),
        title,
        authors,
        journal,
        volume,
        issue,
        pages,
        year,
    });
    applyLocator(reference, locator);
    return reference;
}

function parseAuthors(segment: string): string[] {
    return splitAuthorsDelimiters(segment, /\s+and\s+|,/);
}

export const __all__ = ["parseChicagoCitation"];
